File: src/main/resources/Clustering/kmeans_seed.py
from pyspark.sql import SparkSession
from pyspark.ml.clustering import KMeans

spark = SparkSession.builder.appName('kmeans').getOrCreate()
# Load data set
dataset = spark.read.csv("seeds_dataset.csv",header=True,inferSchema=True)
dataset.head(1)
# Show statistics
dataset.describe().show()
# Format the data
from pyspark.ml.feature import VectorAssembler
dataset.columns
assembler = VectorAssembler(inputCols= dataset.columns,outputCol='features')
final_data = assembler.transform(dataset)
final_data.printSchema()
# Scale the data
# It is good to scale our data to deal with the curse of dimensionality
from pyspark.ml.feature import StandardScaler
scaler = StandardScaler(inputCol="features",outputCol="scaledFeatures")
# Compute summary statistics by fitting StandardScaler
scalerModel = scaler.fit(final_data)
# Normalize each feature to hae unit standard deviation
final_data = scalerModel.transform(final_data)
# Train the model and evaluate
kmeans = KMeans(featuresCol='scaledFeatures', k=3)
model = kmeans.fit(final_data)
# Shows the results
centers = model.clusterCenters()
print("Cluster Centers: ")
for center in centers:
    print(center)

model.transform(final_data).select('prediction').show()

File: src/main/scala/com/normanlimxk/SparkML/Clustering/KmeansSeed.scala
package com.normanlimxk.SparkML.Clustering

import com.normanlimxk.SparkML.SparkSessionWrapper
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.feature.VectorAssembler
import org.apache.spark.ml.feature.StandardScaler
import org.apache.spark.ml.clustering.KMeans

object KmeansSeed extends SparkSessionWrapper {
  def main(args: Array[String]): Unit = {

    // Load data set
    val dataset = spark.read
      .option("header", "true")
      .option("inferSchema", "true")
      .csv("src/main/resources/Clustering/seeds_dataset.csv")
    dataset.show(1, false)

    // Show statistics
    dataset.describe().show()

    // Format the data
    dataset.columns.foreach(println)
    val vecAssember = new VectorAssembler()
      .setInputCols(dataset.columns)
      .setOutputCol("features")
    val vectorizedData = vecAssember.transform(dataset)

    // Scale the data to deal with the curse of dimensionality
    // StandardScaler transforms a dataset of Vector rows, normalizing each feature to have unit standard deviation and/or zero mean.
    val scaler = new StandardScaler()
      .setInputCol("features")
      .setOutputCol("scaledFeatures")
      .setWithStd(true) // scale the data to unit standard deviation
      .setWithMean(false) // do not center the mean

    // Compute summary statistics by fitting StandardScaler
    val scalerModel = scaler.fit(vectorizedData)

    // Normalize each feature to have unit standard deviation
    val normalizedData = scalerModel.transform(vectorizedData)

    // Train the model and evaluate
    // predefined no of clusters is 3
    val kmeans = new KMeans()
      .setFeaturesCol("scaledFeatures")
      .setK(3)
    val model = kmeans.fit(normalizedData)

    // Show results
    val centers = model.clusterCenters
    println("Cluster Centers:")
    centers.foreach(println)

    model.transform(normalizedData).select("prediction").show(truncate = false)
  }
}
